### PR TITLE
CDAP-21004: SCM migration service

### PR DESCRIPTION
**Pull request by samdgupi**
Added a new service that runs periodically to migrate existing metadata from `ApplicationSpecification` table to `NamespaceSourceControlMetadata` table. 

This is not added as part of startup tasks to avoid blocking the system startup for a non-critical feature. As the scm metadata is considered to be eventually consistent anyway.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (itsankit-google) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:45]: Do we really need to run it on schedule?

Can it be a `AbstractIdleService` or `AbstractExecutionThreadService` instead?

Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:45]: We can run it only once per service start. The only downside is if the thread stops because of any uncaught exception (maybe DB is unreachable) we need to wait till next restart for the data to be populated. 
We also do not want to block on this service to be complete running for the service to start as SCM data is anyway should be eventually consistent
@tivv Do you have any suggestion here ?

Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:45]: Modified it so that in case success it stops itself. Otherwise retries.
@tivv @itsankit-google Please check and let me know if this makes sense

Comment (itsankit-google) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:45]: SGTM, will wait for @tivv's approval.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (adrikagupta) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java]: Please change it to
`private static final Logger LOG = LoggerFactory.getLogger(SourceControlMetadataMigrationService.class);`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (adrikagupta) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:117]: Here, `setAppSourceControlMeta` method sets the sync status to `true`. We can't say the sync status will be true for sure. Do you think during migration we should set it as false and periodic refresh service will compute it's correct value?

File: cdap-common/src/main/resources/cdap-default.xml
Comment (adrikagupta) [inline cdap-common/src/main/resources/cdap-default.xml]: We have the capability that if something goes wrong during the migration, it will retry and stop once it is successful. So, shouldn't we have a lesser interval time, say 5 minutes? If anything goes wrong during migration, user would have to wait for 12 hours which might be a lot.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:117]: We should default to `True` as we should treat the last sync data to be source of truth. We can think this migration as a bunch of pull/push happened together

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java]: nit: update the comment

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java]: I'd recommend newFixedDelaySchedule to ensure there is always delay between runs. Also you can put 0 as first run delay.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:112]: In which case would it be null? I see it's read from GSON. It must be a very special case for it to produce null.

Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java:112]: In ApplicationSepcification table we only added SCM meta when push/pulled happened. Hence this value can be null when the meta was never there in the old table. GSON.fromJson auto handles input null value

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java
Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataMigrationService.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java
Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/store/Store.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java:202]: Please add to "shutdown" as well

**Review by tivv — APPROVED** [review]
LGTM for service usage with small comment

**Review by adrikagupta — APPROVED** [review]

**Pull request merged**
